src/runtime-rs/crates/hypervisor/src/qemu/cmdline_generator.rs

```diff
@@ -1178,6 +1181,82 @@ impl ToQemuParams for DeviceIntelIommu {
     }
 }
 
+// Qemu provides methods and types for managing QEMU instances.
+// To manage a qemu instance after it has been launched you need
+// to pass the -qmp option during launch requesting the qemu instance
+// to create a QMP unix domain manageent socket, e.g.,
+// -qmp unix:/tmp/qmp-socket,server=on,wait=off.
+#[derive(Debug, Default, PartialEq)]
+#[allow(dead_code)]
+pub enum MonitorProtocol {
+    // Socket using a human-friendly text-based protocol.
+    Hmp,
+
+    // Socket using a richer json-based protocol.
+    #[default]
+    Qmp,
+
+    // Same as Qmp with pretty json formatting.
+    QmpPretty,
+}
+
+#[allow(dead_code)]
+impl MonitorProtocol {
+    pub fn new(proto: &str) -> Self {
+        match proto {
+            "hmp" => MonitorProtocol::Hmp,
+            "qmp-pretty" => MonitorProtocol::QmpPretty,
+            _ => MonitorProtocol::Qmp,
+        }
+    }
+}
+
+impl ToString for MonitorProtocol {
+    fn to_string(&self) -> String {
+        match *self {
+            MonitorProtocol::Hmp => "hmp".to_string(),
+            MonitorProtocol::QmpPretty => "qmp-pretty".to_string(),
+            _ => "qmp".to_string(),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct QmpSocket {
+    // type is the socket type (e.g. "unix").
+    pub socket_type: String,
+    // protocol to be used on the socket.
+    pub protocol: MonitorProtocol,
+    // QMP listener file descriptor to be passed to qemu
+    pub fd: Option<File>,
+    // name is the socket name.
+    pub name: String,
```

Comment:
Yes, I don't think we'll ever use both of these options, will we?  We should leave out the one that we don't use.  The rationale is again, unlike (the original) govmm our ambition is not a generic qemu support that can do whatever qemu can do - for the sake of simplicity, we only support what's actually useful for runtime-rs.

Reply:
> IMO, we can just use the `unix:fd=xxx..` and no need to keep the `name` for `unix:path=xx...`. And I prepare to remove the field `name` in the next version.

Hold on. It is true that the primary QMP control point will only be `unix:fd=XXX` but the go runtime also offers the `extra_monitor_socket` configuration that allows to add an extra QMP/HMP control point for debugging purpose. The latter can only be `unix:path=xx...`.

Reply:
Right, that's a good point.  We need to decide now whether the base and extra sockets share enough configuration to justify both being represented by the same struct - otherwise the extra socket might be better represented by a separate struct.